src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt

```diff
@@ -0,0 +1,369 @@
+                val parameter = Cypher.parameter(variablePrefix + "_" + objectField.name)
+                val condition = predicate.conditionCreator(prop, parameter)
+                filterParams[parameter.name] = objectField.value.toJavaValue()
+                where = conditionAdder(where, condition)
+            }
+            return where
+        }
+
+
+        private fun createRelation(rel: RelationshipInfo, start: Node, end: Node): Relationship =
+                when (rel.out) {
+                    false -> start.relationshipFrom(end, rel.relType)
+                    true -> start.relationshipTo(end, rel.relType)
+                    null -> start.relationshipBetween(end, rel.relType)
+                }
+    }
+
+    companion object {
+        private val LOGGER = LoggerFactory.getLogger(OptimizedFilterHandler::class.java)
+
+        data class Predicate<T>(
+                val op: T,
+                val queryField: ObjectField,
+                val fieldDefinition: GraphQLFieldDefinition,
+                val index: Int)
+
+        class ParsedQuery(objectValue: ObjectValue, type: GraphQLFieldsContainer) {
+            val fieldPredicates = mutableListOf<Predicate<FieldOperator>>()
+            val relationPredicates = mutableListOf<Predicate<RelationOperator>>()
+            val or: List<Value<*>>?
+            val and: List<Value<*>>?
+
+            init {
+                val queriedFields = objectValue.objectFields
+                    .mapIndexed { index, field -> field.name to (index to field) }
+                    .toMap(mutableMapOf())
+                or = queriedFields.remove("OR")?.second?.value?.let {
+                    return@let (it as ArrayValue).values
+                            ?: throw IllegalArgumentException("OR is expected to be a list")
+                }
+                and = queriedFields.remove("AND")?.second?.value?.let {
+                    return@let (it as ArrayValue).values
+                            ?: throw IllegalArgumentException("AND is expected to be a list")
+                }
+                for (definedField in type.fieldDefinitions) {
+                    if (definedField.isRelationship()) {
+                        collectRelationPredicates(type, definedField, queriedFields)
+                    } else {
+                        collectFieldPredicates(definedField, queriedFields)
+                    }
+                }
+                if (queriedFields.isNotEmpty()) {
+                    throw OptimizedQueryException("queried unknown fields: " + queriedFields.keys)
+                }
+                fieldPredicates.sortedBy(Predicate<FieldOperator>::index)
+                relationPredicates.sortedBy(Predicate<RelationOperator>::index)
+            }
+
+            private fun collectRelationPredicates(type: GraphQLFieldsContainer, definedField: GraphQLFieldDefinition, queriedFields: MutableMap<String, Pair<Int, ObjectField>>) {
```

Comment:
this is a pretty huge method, can we split it up,  and perhaps move stuff into the Operator?

Reply:
moved into `RelationOperator`